File: leetcode3/이진희/860. Lemonade Change.java
/* 

1. 아이디어 : 거스름돈을 줄 수 있는지 체크 
            10$ -> 5$ 1장
            20$ -> 5$ 3장 or 10$ + 5$ 

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : 완전탐색, 조건문 

 */

class Solution {
    public boolean lemonadeChange(int[] bills) {

        // 5 10
        int[] money = new int[2];

        for(int i=0; i<bills.length; i++) {
           int pay = bills[i];

           if(pay == 5) money[0]++;
           else if(pay == 10) {
                if(money[0] == 0) return false;
                money[1]++;
                money[0]--;
           }
           else {
                // 5$ 3장 or 10$ + 5$
                if(money[0]>0 && money[1]>0) {
                    money[0]--;
                    money[1]--;
                }
                else if (money[0] > 2) money[0] -=3;
                else return false;
           }
        }

        return true;
    }
}

File: leetcode3/최원준/860. Lemonade Change.py
#

'''
1. 아이디어 :
-

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
dict

'''
class Solution:
    def lemonadeChange(self, bills: list[int]) -> bool:
        count = {5:0, 10:0, 20:0} #5, 10, 20

        def is_possible(payment):
            if payment == 20:
                if count[10] >= 1 and count[5] >= 1:
                    count[10] -= 1
                    count[5] -= 1
                    count[20] += 1
                    return True
                elif count[10] == 0 and count[5] >= 3:
                    count[5] -= 3
                    count[20] += 1
                    return True
            elif payment == 10:
                if count[5] >= 1:
                    count[5] -= 1
                    count[10] += 1
                    return True
            elif payment == 5:
                count[5] += 1
                return True
            
            return False

        for b in bills:
            if not is_possible(b):
                return False
        return True
                    
